components/mailchimp/sources/new-or-updated-list-segment/new-or-updated-list-segment.js

```diff
@@ -0,0 +1,129 @@
+    db: "$.service.db",
+  },
+  hooks: {
+    async deploy() {
+      // Emits sample events on the first run during deploy.
+      const mailchimpAudienceSegmentsInfo =
+        await this.mailchimp.getAudienceSegments(
+          this.server,
+          this.listId,
+          10,
+          0,
+          null,
+          null,
+          null,
+          null
+        );
+      const { segments: mailchimpAudienceSegments = [] } =
+        mailchimpAudienceSegmentsInfo;
+      if (!mailchimpAudienceSegments.length) {
+        console.log("No data available, skipping iteration");
+        return;
+      }
+      const relevantDate = ["Created"].includes(this.watchFor)
+        ? mailchimpAudienceSegments[0].created_at
+        : mailchimpAudienceSegments[0].updated_at;
+      this.db.set("lastRelevantDate", relevantDate);
+      mailchimpAudienceSegments.forEach(this.emitEvent);
+    },
+  },
+  methods: {
+    ...common.methods,
+    generateMeta(eventPayload) {
+      if (["Created"].includes(this.watchFor)) {
+        return {
+          id: eventPayload.id,
+          summary: `A new segment "${eventPayload.name}" was created.`,
+          ts: +new Date(eventPayload.created_at),
+        };
+      } else {
+        const ts = +new Date(eventPayload.updated_at);
+        return {
+          id: `${eventPayload.id}${ts}`,
+          summary: `Segment "${eventPayload.name}" was updated.`,
+          ts,
+        };
+      }
+    },
+    emitEvent(eventPayload) {
+      const meta = this.generateMeta(eventPayload);
+      this.$emit(eventPayload, meta);
+    },
+  },
+  async run() {
+    let sinceCreatedAt;
+    let beforeCreatedAt;
+    let sinceUpdatedAt;
+    let beforeUpdatedAt;
+    if (["Created"].includes(this.watchFor)) {
+      sinceCreatedAt = this.db.get("lastRelevantDate");
+      beforeCreatedAt = moment().toISOString();
+    } else {
+      sinceUpdatedAt = this.db.get("lastRelevantDate");
+      beforeUpdatedAt = moment().toISOString();
+    }
+    let mailchimpAudienceSegmentsInfo;
+    let mailchimpAudienceSegments;
+    let offset = 0;
+    do {
+      mailchimpAudienceSegmentsInfo = await this.mailchimp.getAudienceSegments(
+        this.server,
+        this.listId,
+        1000,
+        offset,
+        sinceCreatedAt,
+        beforeCreatedAt,
+        sinceUpdatedAt,
+        beforeUpdatedAt
+      );
+      mailchimpAudienceSegments = mailchimpAudienceSegmentsInfo.segments;
+      if (!mailchimpAudienceSegments.length) {
+        console.log("No data available, skipping iteration");
+        return;
+      }
+      mailchimpAudienceSegments.forEach(this.emitEvent);
+      const relevantDate = ["Created"].includes(this.watchFor)
+        ? mailchimpAudienceSegments[0].created_at
+        : mailchimpAudienceSegments[0].updated_at;
+      this.db.set("lastRelevantDate", relevantDate);
+      offset = offset + mailchimpAudienceSegments.length;
+    } while (mailchimpAudienceSegments.length > 0);
+  },
+};
```

Comment:
We should refactor this to reuse the logic suggested in `new-campaign` (in terms of "sent" vs "created")

Reply:
This has been addressed.